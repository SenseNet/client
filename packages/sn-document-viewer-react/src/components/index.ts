export * from './comment'
export * from './document-widgets'
export * from './page-widgets'
export * from './ConfirmationDialog'
export * from './DocumentViewer'
export * from './DocumentViewerError'
export * from './DocumentViewerLayout'
export * from './DocumentViewerLoading'
export * from './DocumentViewerRegeneratePreviews'
export * from './LayoutAppBar'
export * from './Page'
export * from './PageList'
export * from './styles'
