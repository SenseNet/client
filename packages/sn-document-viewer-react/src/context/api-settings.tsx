import React, { useEffect, useState } from 'react'
import { useRepository } from '@sensenet/hooks-react'
import { File as SnFile } from '@sensenet/default-content-types'
import { deepMerge, toNumber } from '@sensenet/client-utils'
import { v1 } from 'uuid'
import { Repository } from '@sensenet/client-core'
import {
  Annotation,
  CommentData,
  CommentWithoutCreatedByAndId,
  DocumentData,
  DocumentViewerApiSettings,
  Highlight,
  PreviewImageData,
  Redaction,
  Shape,
} from '../models'

/**
 * Adds a globally unique ID to the shape
 */
const addGuidToShape = <T extends Shape>(shape: T) => {
  shape.guid = v1()
  return shape
}

/**
 * Concatenates avatar url with host
 * @param {DocumentData} documentData
 */
function changeCreatedByUrlToCurrent(documentData: DocumentData): (value: CommentData) => CommentData {
  return (comment) => {
    return {
      ...comment,
      createdBy: { ...comment.createdBy, avatarUrl: `${documentData.hostName}${comment.createdBy.avatarUrl}` },
    }
  }
}

export const DocumentViewerApiSettingsContext = React.createContext<DocumentViewerApiSettings>({
  canEditDocument: async () => false,
  canHideRedaction: async () => false,
  canHideWatermark: async () => false,
  commentActions: {
    addPreviewComment: async ({ comment }) => ({ ...comment } as any),
    deletePreviewComment: async () => ({ modified: false }),
    getPreviewComments: async () => [],
  },
  getDocumentData: async () => ({
    documentName: 'example',
    documentType: '',
    fileSizekB: 3,
    hostName: '',
    idOrPath: 1,
    pageAttributes: [],
    pageCount: 0,
    shapes: { annotations: [], highlights: [], redactions: [] },
  }),
  getExistingPreviewImages: async () => [],
  isPreviewAvailable: async () => ({ Height: 0, Width: 0, Index: 0, Attributes: { degree: 0 } }),
  regeneratePreviews: async () => undefined,
  saveChanges: async () => undefined,
})

export const createDefaultApiSettings: (repo: Repository) => DocumentViewerApiSettings = (repo: Repository) => ({
  regeneratePreviews: async ({ document, abortController }) => {
    repo.executeAction({
      idOrPath: document.idOrPath,
      method: 'POST',
      body: undefined,
      name: 'RegeneratePreviews',
      requestInit: {
        signal: abortController.signal,
      },
    })
  },
  saveChanges: async ({ document, pages, abortController }) => {
    const reqBody = {
      Shapes: JSON.stringify([
        { redactions: document.shapes.redactions },
        { highlights: document.shapes.highlights },
        { annotations: document.shapes.annotations },
      ]),
      PageAttributes: JSON.stringify(
        pages
          .map((p) => (p.Attributes && p.Attributes.degree && { pageNum: p.Index, options: p.Attributes }) || undefined)
          .filter((p) => p !== undefined),
      ),
    }
    await repo.patch<SnFile>({
      idOrPath: document.idOrPath,
      content: reqBody,
      requestInit: {
        signal: abortController.signal,
      },
    })
  },
  getDocumentData: async (settings) => {
    const documentData = (
      await repo.load<SnFile>({
        idOrPath: settings.idOrPath,
        oDataOptions: {
          select: 'all',
        },
      })
    ).d

    return {
      idOrPath: settings.idOrPath,
      hostName: repo.configuration.repositoryUrl,
      fileSizekB: documentData.Size as number,
      pageCount: toNumber(documentData.PageCount, -1)!,
      documentName: documentData.DisplayName || '',
      documentType: documentData.Type || 'File',
      shapes: (documentData.Shapes && {
        redactions: (JSON.parse(documentData.Shapes)[0].redactions as Redaction[]).map((a) => addGuidToShape(a)) || [],
        annotations:
          (JSON.parse(documentData.Shapes)[2].annotations as Annotation[]).map((a) => addGuidToShape(a)) || [],
        highlights: (JSON.parse(documentData.Shapes)[1].highlights as Highlight[]).map((a) => addGuidToShape(a)) || [],
      }) || {
        redactions: [],
        annotations: [],
        highlights: [],
      },
      pageAttributes: (documentData.PageAttributes && JSON.parse(documentData.PageAttributes)) || [],
    }
  },
  isPreviewAvailable: async ({ document, version, page, abortController }) => {
    const responseBody = await repo.executeAction<{ page: number }, PreviewImageData & { PreviewAvailable: string }>({
      idOrPath: document.idOrPath,
      method: 'POST',
      name: `PreviewAvailable?version=${version}`,
      body: {
        page,
      },
      requestInit: {
        signal: abortController.signal,
      },
    })

    if (responseBody.PreviewAvailable) {
      responseBody.PreviewImageUrl = `${document.hostName}${responseBody.PreviewAvailable}`
      responseBody.ThumbnailImageUrl = `${document.hostName}${responseBody.PreviewAvailable.replace(
        'preview',
        'thumbnail',
      )}`
      return responseBody as PreviewImageData
    }
  },
  canEditDocument: async ({ document, abortController }) => {
    const response = await repo.security.hasPermission(document.idOrPath, ['Save'], undefined, {
      signal: abortController.signal,
      credentials: 'include',
    })
    return response
  },
  canHideRedaction: async ({ document, abortController }) =>
    await repo.security.hasPermission(document.idOrPath, ['PreviewWithoutRedaction'], undefined, {
      signal: abortController.signal,
      credentials: 'include',
    }),
  canHideWatermark: async ({ document, abortController }) =>
    await repo.security.hasPermission(document.idOrPath, ['PreviewWithoutWatermark'], undefined, {
      signal: abortController.signal,
      credentials: 'include',
    }),
  getExistingPreviewImages: async ({ document, version, abortController }) => {
    if (document.pageCount < -1) {
      throw Error('Preview generation error')
    }

    const response = await repo.executeAction({
      idOrPath: document.idOrPath,
      name: `GetExistingPreviewImages`,
      method: 'POST',
      body: {},
      oDataOptions: {
        select: 'all',
        expand: 'all',
        version,
      } as any,
      requestInit: {
        signal: abortController.signal,
      },
    })

    const availablePreviews = ((await response) as Array<PreviewImageData & { PreviewAvailable?: string }>).map((a) => {
      if (a.PreviewAvailable) {
        a.PreviewImageUrl = `${document.hostName}${a.PreviewAvailable}`
        a.ThumbnailImageUrl = `${document.hostName}${a.PreviewAvailable.replace('preview', 'thumbnail')}`
      }
      return a
    })

    const allPreviews: PreviewImageData[] = []
    for (let i = 0; i < document.pageCount; i++) {
      allPreviews[i] = availablePreviews[i] || ({ Index: i + 1 } as any)
      const pageAttributes = document.pageAttributes.find((p) => p.pageNum === allPreviews[i].Index)
      allPreviews[i].Attributes = pageAttributes && pageAttributes.options
    }
    return allPreviews
  },
  commentActions: {
    addPreviewComment: async ({ document, comment, abortController }) => {
      const response = await repo.executeAction<CommentWithoutCreatedByAndId, CommentData>({
        idOrPath: document.idOrPath,
        body: comment,
        name: 'AddPreviewComment',
        method: 'POST',
        requestInit: {
          signal: abortController.signal,
        },
      })
      return [response].map(changeCreatedByUrlToCurrent(document))[0]
    },
    deletePreviewComment: async ({ document, commentId, abortController }) => {
      const response = await repo.executeAction<any, { modified: boolean }>({
        idOrPath: document.idOrPath,
        body: { id: commentId },
        name: 'DeletePreviewComment',
        method: 'POST',
        requestInit: {
          signal: abortController.signal,
        },
      })
      return response
    },
    getPreviewComments: async ({ document, page, abortController }) => {
      const response = await repo.executeAction<any, CommentData[]>({
        idOrPath: document.idOrPath,
        name: 'GetPreviewComments',
        method: 'GET',
        oDataOptions: {
          page,
        } as any,
        requestInit: {
          signal: abortController.signal,
        },
      })
      return response.map(changeCreatedByUrlToCurrent(document))
    },
  },
})

export const DocumentViewerApiSettingsProvider: React.FC<{ options?: Partial<DocumentViewerApiSettings> }> = ({
  children,
  options,
}) => {
  const repo = useRepository()
  const [settings, setSettings] = useState(deepMerge(createDefaultApiSettings(repo), options))
  useEffect(() => {
    setSettings(deepMerge(createDefaultApiSettings(repo), options))
  }, [repo, options])

  return (
    <DocumentViewerApiSettingsContext.Provider value={settings}>{children}</DocumentViewerApiSettingsContext.Provider>
  )
}
