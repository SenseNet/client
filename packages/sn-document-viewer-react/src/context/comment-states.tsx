import React, { useState } from 'react'
import { DraftCommentMarker } from '@sensenet/client-core'

export interface CommentState {
  draft?: DraftCommentMarker
  setDraft: (d?: DraftCommentMarker) => void
  activeCommentId?: string
  setActiveComment: (commentId?: string) => void
}

export const defaultCommentState: CommentState = {
  setDraft: () => undefined,
  activeCommentId: undefined,
  setActiveComment: () => undefined,
}

export const CommentStateContext = React.createContext<CommentState>(defaultCommentState)

export const CommentStateProvider: React.FC = ({ children }) => {
  const [draft, setDraft] = useState<DraftCommentMarker>()
  const [activeCommentId, setActiveComment] = useState<string>()

  return (
    <CommentStateContext.Provider value={{ draft, setDraft, activeCommentId, setActiveComment }}>
      {children}
    </CommentStateContext.Provider>
  )
}
