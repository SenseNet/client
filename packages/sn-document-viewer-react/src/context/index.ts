export * from './api-settings'
export * from './comment-states'
export * from './comments'
export * from './document-data'
export * from './document-permissions'
export * from './localization-context'
export * from './preview-image-data'
export * from './viewer-settings'
export * from './viewer-state'
