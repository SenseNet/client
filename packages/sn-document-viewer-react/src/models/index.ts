export * from './DocumentViewerApiSettings'
export * from './Theming'
export * from './viewer-state'
