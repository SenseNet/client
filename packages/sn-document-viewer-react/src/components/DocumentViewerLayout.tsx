import Drawer from '@material-ui/core/Drawer'
import { SlideProps } from '@material-ui/core/Slide'
import Typography from '@material-ui/core/Typography'
import React = require('react')
import { connect } from 'react-redux'
import { DraftCommentMarker } from '../models'
import { componentType } from '../services'
import { createComment, RootReducerType, setActivePages, setThumbnails } from '../store'
import {
  getComments,
  setSelectedCommentId,
  toggleIsCreateCommentActive,
  toggleIsPlacingCommentMarker,
} from '../store/Comments'
import Comment from './comment/Comment'
import { CreateComment } from './comment/CreateComment'
import { CommentsContainer, PageList } from './'

/**
 * maps state fields from the store to component props
 * @param state the redux state
 */
const mapStateToProps = (state: RootReducerType) => {
  return {
    activePages: state.sensenetDocumentViewer.viewer.activePages,
    zoomMode: state.sensenetDocumentViewer.viewer.zoomMode,
    customZoomLevel: state.sensenetDocumentViewer.viewer.customZoomLevel,
    showThumbnails: state.sensenetDocumentViewer.viewer.showThumbnails,
    showComments: state.sensenetDocumentViewer.viewer.showComments,
    comments: state.comments.items,
    isCreateCommentActive: state.comments.isCreateCommentActive,
    isPlacingCommentMarker: state.comments.isPlacingCommentMarker,
    selectedCommentId: state.comments.selectedCommentId,
    fitRelativeZoomLevel: state.sensenetDocumentViewer.viewer.fitRelativeZoomLevel,
    localization: state.sensenetDocumentViewer.localization,
  }
}

/**
 * maps state actions from the store to component props
 * @param state the redux state
 */
const mapDispatchToProps = {
  setActivePages,
  getComments,
  setThumbnails,
  createComment,
  setSelectedCommentId,
  toggleIsCreateCommentActive,
  toggleIsPlacingCommentMarker,
}

/** Props definition for the Document Viewer layout */
export interface DocumentLayoutOwnProps {
  drawerSlideProps?: Partial<SlideProps>
}

/** State type definition for the DocumentViewerLayout component */
export interface DocumentLayoutState {
  draftCommentMarker?: DraftCommentMarker
  activePage?: number
  thumbnaislVisibility: boolean
  createCommentValue: string
}

/**
 * Component for the main DocumentViewer layout
 */
export class DocumentViewerLayoutComponent extends React.Component<
  componentType<typeof mapStateToProps, typeof mapDispatchToProps, DocumentLayoutOwnProps>,
  DocumentLayoutState
> {
  constructor(props: DocumentViewerLayoutComponent['props']) {
    super(props)
    this.state = {
      activePage: 1,
      thumbnaislVisibility: this.props.showThumbnails,
      createCommentValue: '',
    }
    this.commentsContainerRef = React.createRef()
    this.createComment = this.createComment.bind(this)
    this.handleKeyUp = this.handleKeyUp.bind(this)
    this.handleMarkerCreation = this.handleMarkerCreation.bind(this)
  }

  private commentsContainerRef: React.RefObject<HTMLDivElement>

  public componentDidMount() {
    document.addEventListener('keyup', this.handleKeyUp)
  }

  public componentWillUnmount() {
    document.removeEventListener('keyup', this.handleKeyUp)
  }

  /** scrolls the viewer to focus to the page with the provided index */
  public scrollTo(index: number, smoothScroll: boolean = true) {
    this.setState({ ...this.state, activePage: index }, () => {
      const pagesContainer = document.getElementById('sn-document-viewer-pages')
      const activePage = document.getElementById(`Page-${index}`)

      if (pagesContainer && pagesContainer.scrollTo && activePage) {
        pagesContainer.scrollTo({
          top: activePage.offsetTop - 8,
          behavior: smoothScroll ? 'smooth' : 'auto',
        })
      }

      const thumbnailsContainer = document.getElementById('sn-document-viewer-thumbnails')
      const activeThumbnail = document.getElementById(`Thumbnail-${index}`)

      if (thumbnailsContainer && thumbnailsContainer.scrollTo && activeThumbnail) {
        thumbnailsContainer.scrollTo({
          top: activeThumbnail.offsetTop - 16,
          behavior: 'smooth',
        })
      }

      if (this.props.activePages[0] !== index) {
        this.props.setActivePages([index])
        this.props.getComments()
      }
    })
  }

  private handleMarkerCreation(draftCommentMarker: DraftCommentMarker) {
    this.setState({ ...this.state, draftCommentMarker })
  }

  private createComment(text: string) {
    if (!this.state.draftCommentMarker || !this.state.activePage) {
      return
    }
    this.props.createComment({ page: this.state.activePage, text, ...this.state.draftCommentMarker })
    this.toggleIsPlacingCommentMarker(false)
  }

  private toggleIsPlacingCommentMarker(isPlacingCommentMarker = !this.props.isPlacingCommentMarker) {
    this.props.toggleIsPlacingCommentMarker(isPlacingCommentMarker)
    this.setState({ ...this.state, draftCommentMarker: undefined })
  }

  private handleKeyUp(ev: KeyboardEvent) {
    if (ev.key !== 'Escape') {
      return
    }
    if (this.props.isPlacingCommentMarker) {
      return this.toggleIsPlacingCommentMarker(false)
    }
    if (this.props.isCreateCommentActive) {
      this.props.toggleIsCreateCommentActive(false)
    }
    this.props.setSelectedCommentId('')
    this.setState({ ...this.state, createCommentValue: '' })
  }

  /** triggered when the component will receive props */
  public componentWillReceiveProps(newProps: this['props']) {
    if (
      this.props.activePages[0] !== newProps.activePages[0] ||
      this.props.fitRelativeZoomLevel !== newProps.fitRelativeZoomLevel
    ) {
      this.scrollTo(newProps.activePages[0], this.props.fitRelativeZoomLevel === newProps.fitRelativeZoomLevel)
      this.toggleIsPlacingCommentMarker(false)
    }
    if (this.props.showThumbnails !== newProps.showThumbnails) {
      if (newProps.showThumbnails) {
        this.setState({
          ...this.state,
          thumbnaislVisibility: true,
        })
        window.dispatchEvent(new Event('resize'))
      } else {
        setTimeout(() => {
          this.setState({
            ...this.state,
            thumbnaislVisibility: false,
          })
          window.dispatchEvent(new Event('resize'))
        }, 200)
      }
    }
    if (this.props.selectedCommentId !== newProps.selectedCommentId) {
      const selectedCommentNode = document.getElementById(newProps.selectedCommentId)
      if (!selectedCommentNode) {
        return
      }
      this.commentsContainerRef.current &&
        this.commentsContainerRef.current.scrollTo({
          top: selectedCommentNode.offsetTop - selectedCommentNode.scrollHeight,
          behavior: 'smooth',
        })
    }
  }

  /**
   * renders the component
   */
  public render() {
    return (
      <div
        style={{
          width: '100%',
          height: '100%',
          display: 'flex',
          flexDirection: 'column',
        }}>
        {this.props.children}
        <div
          style={{
            display: 'flex',
            height: '100%',
            width: '100%',
            overflow: 'hidden',
            zIndex: 0,
            position: 'relative',
          }}>
          <Drawer
            variant={'persistent'}
            open={this.props.showThumbnails}
            anchor="left"
            SlideProps={this.props.drawerSlideProps}
            PaperProps={{
              style: {
                position: 'relative',
                width: this.state.thumbnaislVisibility ? '200px' : 0,
                height: '100%',
                overflow: 'hidden',
              },
            }}>
            <PageList
              showWidgets={false}
              style={{ minWidth: 200, marginRight: '-16px', paddingRight: 0 }}
              id="sn-document-viewer-thumbnails"
              zoomMode="fit"
              fitRelativeZoomLevel={0}
              zoomLevel={1}
              onPageClick={(_ev, index) => this.scrollTo(index)}
              elementNamePrefix="Thumbnail-"
              images="thumbnail"
              tolerance={0}
              padding={16}
              activePage={this.state.activePage}
            />
          </Drawer>
          <PageList
            handleMarkerCreation={this.handleMarkerCreation}
            showWidgets={true}
            id="sn-document-viewer-pages"
            zoomMode={this.props.zoomMode}
            zoomLevel={this.props.customZoomLevel}
            fitRelativeZoomLevel={this.props.fitRelativeZoomLevel}
            onPageClick={(_ev, index) => this.scrollTo(index)}
            elementNamePrefix="Page-"
            images="preview"
            tolerance={0}
            padding={8}
            activePage={this.state.activePage}
          />
          <Drawer
            variant={'persistent'}
            open={this.props.showComments}
            anchor="right"
            SlideProps={this.props.drawerSlideProps}
            PaperProps={{
              style: {
                position: 'relative',
                width: this.props.showComments ? '340px' : 0,
                height: '100%',
                overflow: 'hidden',
              },
            }}>
            <CommentsContainer ref={this.commentsContainerRef}>
              <Typography variant="h4">{this.props.localization.commentSideBarTitle}</Typography>
              <CreateComment
                isActive={this.props.isCreateCommentActive}
                handleIsActive={isActive => this.props.toggleIsCreateCommentActive(isActive)}
                draftCommentMarker={this.state.draftCommentMarker}
                handlePlaceMarkerClick={isPlacing => this.toggleIsPlacingCommentMarker(isPlacing)}
                isPlacingMarker={this.props.isPlacingCommentMarker}
                localization={this.props.localization}
                createComment={this.createComment}
                inputValue={this.state.createCommentValue}
                handleInputValueChange={value => this.setState({ ...this.state, createCommentValue: value })}
              />
              {this.props.comments.map(comment => (
                <Comment key={comment.id} {...comment} />
              ))}
            </CommentsContainer>
          </Drawer>
        </div>
      </div>
    )
  }
}

const connectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps,
)(DocumentViewerLayoutComponent)

export { connectedComponent as DocumentViewerLayout }
