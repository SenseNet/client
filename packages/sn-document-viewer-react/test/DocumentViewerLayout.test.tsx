import { ObservableValue } from '@sensenet/client-utils'
import { mount, shallow } from 'enzyme'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { ThemeProvider } from 'styled-components'
import { Comment, CommentsContainer, CreateComment, PageList } from '../src/components'
import { DocumentViewerLayout } from '../src/components/DocumentViewerLayout'
import { DocumentViewerApiSettingsContext } from '../src/context/api-settings'
import { CommentStateContext, defaultCommentState } from '../src/context/comment-states'
import { CommentsContext, defaultCommentsContext } from '../src/context/comments'
import { defaultViewerState, ViewerStateContext } from '../src/context/viewer-state'
import { defaultTheme } from '../src/models/Theming'
import { PreviewImageDataContextWrapper } from './__Mocks__/preview-image-data-context-wrapper'
import { ViewerStateContextWrapper } from './__Mocks__/viewer-state-context-wrapper'
import { defaultSettings, examplePreviewComment } from './__Mocks__/viewercontext'

declare global {
  interface Window {
    domNode: HTMLDivElement
  }
}

describe('Document Viewer Layout component', () => {
  beforeEach(() => {
    const div = document.createElement('div')
    window.domNode = div
    document.body.appendChild(div)
  })

  it('should scroll to page when page changed', () => {
    const onPageChange = new ObservableValue(1)
    const scrollToMock = jest.fn()
    ;(window as any).HTMLElement.prototype.scrollTo = scrollToMock
    mount(
      <PreviewImageDataContextWrapper>
        <DocumentViewerLayout pagePadding={0} thumbnailPadding={0}>
          {'some children'}
        </DocumentViewerLayout>
      </PreviewImageDataContextWrapper>,
      {
        wrappingComponent: ViewerStateContextWrapper,
        wrappingComponentProps: { onPageChange, showThumbnails: true },
        attachTo: window.domNode,
      },
    )

    onPageChange.setValue(3)
    expect(scrollToMock).toBeCalled()
  })

  it('click on a page should scroll to the selected page', async () => {
    const scrollToMock = jest.fn()
    ;(window as any).HTMLElement.prototype.scrollTo = scrollToMock
    const wrapper = mount(
      <PreviewImageDataContextWrapper>
        <DocumentViewerLayout pagePadding={0} thumbnailPadding={0}>
          {'some children'}
        </DocumentViewerLayout>
      </PreviewImageDataContextWrapper>,
      {
        wrappingComponent: ViewerStateContextWrapper,
        attachTo: window.domNode,
      },
    )

    wrapper.find(PageList).last().prop('onPageClick')({} as any, 3)

    expect(scrollToMock).toBeCalled()
  })

  it('click on a thumbnail should scroll to the selected page and change viewerState', async () => {
    const scrollToMock = jest.fn()
    const updateState = jest.fn()
    ;(window as any).HTMLElement.prototype.scrollTo = scrollToMock
    const wrapper = mount(
      <ViewerStateContext.Provider
        value={{
          ...defaultViewerState,
          updateState,
        }}>
        <PreviewImageDataContextWrapper>
          <DocumentViewerLayout pagePadding={0} thumbnailPadding={0}>
            {'some children'}
          </DocumentViewerLayout>
        </PreviewImageDataContextWrapper>
      </ViewerStateContext.Provider>,
      {
        wrappingComponent: ViewerStateContextWrapper,
        attachTo: window.domNode,
      },
    )

    wrapper.find(PageList).first().prop('onPageClick')({} as any, 3)

    expect(scrollToMock).toBeCalled()
    expect(updateState).toBeCalledWith({ visiblePages: [3] })
  })

  it('should show comments', async () => {
    let wrapper: any
    await act(async () => {
      wrapper = mount(
        <CommentsContext.Provider value={{ ...defaultCommentsContext, comments: [examplePreviewComment] }}>
          <ViewerStateContext.Provider value={{ ...defaultViewerState, showComments: true }}>
            <DocumentViewerLayout>{'some children'}</DocumentViewerLayout>
          </ViewerStateContext.Provider>
        </CommentsContext.Provider>,
      )
    })
    expect(wrapper.find(CommentsContainer).exists()).toBeTruthy()
  })

  it('should handle comment creation', async () => {
    const addPreviewComment = function addPreviewComment() {
      return { id: 'someId2', page: 2, text: 'Thats a comment2', x: 20, y: 20, createdBy: {} }
    }
    let wrapper: any
    await act(async () => {
      wrapper = mount(
        <DocumentViewerApiSettingsContext.Provider
          value={{
            ...defaultSettings,
            commentActions: { ...defaultSettings.commentActions, addPreviewComment: addPreviewComment as any },
          }}>
          <CommentStateContext.Provider value={{ draft: { id: 'id', x: 10, y: 10 } } as any}>
            <ViewerStateContext.Provider value={{ ...defaultViewerState, showComments: true }}>
              <DocumentViewerLayout />
            </ViewerStateContext.Provider>
          </CommentStateContext.Provider>
        </DocumentViewerApiSettingsContext.Provider>,
      )
    })

    const text = 'this is the comment'
    await act(async () => {
      wrapper.find(CreateComment).prop('createComment')(text)
    })

    // There is already a comment this is why length should be 2
    expect(wrapper.update().find(Comment)).toHaveLength(2)
  })

  it.skip('should handle comment input value change', async () => {
    const wrapper = shallow(
      <ViewerStateContext.Provider value={{ ...defaultViewerState, showComments: true }}>
        <DocumentViewerLayout>{'some children'}</DocumentViewerLayout>
      </ViewerStateContext.Provider>,
    )

    const text = 'this is the comment'
    wrapper.find(CreateComment).prop('handleInputValueChange')(text)
    expect(wrapper.state('createCommentValue')).toBe(text)
  })

  it('should handle create comment isActive', () => {
    const updateState = jest.fn()
    const wrapper = mount(<DocumentViewerLayout />, {
      wrappingComponent: ViewerStateContextWrapper,
      wrappingComponentProps: { updateState },
    })

    wrapper.find(CreateComment).prop('handleIsActive')(true)
    expect(updateState).toBeCalledTimes(1)
    expect(updateState).toBeCalledWith({ isCreateCommentActive: true })
  })

  const events: any = {}
  document.addEventListener = jest.fn((event, cb) => {
    events[event] = cb
  })

  it('should handle esc keyup', async () => {
    const setActiveComment = jest.fn()

    mount(
      <ViewerStateContext.Provider value={{ ...defaultViewerState, showComments: true }}>
        <CommentStateContext.Provider value={{ ...defaultCommentState, setActiveComment }}>
          <DocumentViewerLayout>{'some children'}</DocumentViewerLayout>
        </CommentStateContext.Provider>
      </ViewerStateContext.Provider>,
    )

    events.keyup({ key: 'a' })
    expect(setActiveComment).toBeCalledTimes(0)
    await act(async () => {
      events.keyup({ key: 'Escape' })
    })

    expect(setActiveComment).toBeCalledTimes(1)
    expect(setActiveComment).toBeCalledWith(undefined)
  })

  it('should abort commenting when esc was pushed', async () => {
    const updateState = jest.fn()
    mount(
      <ThemeProvider theme={defaultTheme}>
        <ViewerStateContext.Provider
          value={{ ...defaultViewerState, showComments: true, isCreateCommentActive: true, updateState }}>
          <DocumentViewerLayout>{'some children'}</DocumentViewerLayout>
        </ViewerStateContext.Provider>
      </ThemeProvider>,
    )
    await act(async () => {
      events.keyup({ key: 'Escape' })
    })

    expect(updateState).toBeCalledWith({ isCreateCommentActive: false })
  })

  it('should abort marker placement but not commenting when esc was pushed', async () => {
    const updateState = jest.fn()
    mount(
      <ThemeProvider theme={defaultTheme}>
        <ViewerStateContext.Provider
          value={{
            ...defaultViewerState,
            showComments: true,
            isCreateCommentActive: true,
            isPlacingCommentMarker: true,
            updateState,
          }}>
          <DocumentViewerLayout>{'some children'}</DocumentViewerLayout>
        </ViewerStateContext.Provider>
      </ThemeProvider>,
    )
    await act(async () => {
      events.keyup({ key: 'Escape' })
    })
    expect(updateState).toBeCalledWith({ isPlacingCommentMarker: false })
    expect(updateState).not.toBeCalledWith({ isCreateCommentActive: false })
  })

  it.skip('should scroll to comment when selectedCommentId is changed', async () => {
    const scrollToMock = jest.fn()
    ;(window as any).HTMLElement.prototype.scrollTo = scrollToMock
    mount(<DocumentViewerLayout />, { attachTo: window.domNode })
    expect(scrollToMock).toBeCalled()
  })

  it.skip('should not scroll to comment when comment is not found', async () => {
    const scrollToMock = jest.fn()
    ;(window as any).HTMLElement.prototype.scrollTo = scrollToMock
    mount(<DocumentViewerLayout />, { attachTo: window.domNode })

    expect(scrollToMock).not.toBeCalled()
  })
})
