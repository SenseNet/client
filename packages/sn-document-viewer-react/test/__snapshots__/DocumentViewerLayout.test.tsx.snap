// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document Viewer Layout component should render without crashing 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "width": "100%",
    }
  }
>
  some children
  <div
    style={
      Object {
        "display": "flex",
        "height": "100%",
        "overflow": "hidden",
        "position": "relative",
        "width": "100%",
        "zIndex": 0,
      }
    }
  >
    <WithStyles(Drawer)
      PaperProps={
        Object {
          "style": Object {
            "height": "100%",
            "overflow": "hidden",
            "position": "relative",
            "width": "200px",
          },
        }
      }
      anchor="left"
      open={true}
      variant="persistent"
    >
      <Connect(PageListComponent)
        activePage={1}
        elementNamePrefix="Thumbnail-"
        fitRelativeZoomLevel={0}
        id="sn-document-viewer-thumbnails"
        images="thumbnail"
        onPageClick={[Function]}
        padding={16}
        showWidgets={false}
        style={
          Object {
            "marginRight": "-16px",
            "minWidth": 200,
            "paddingRight": 0,
          }
        }
        tolerance={0}
        zoomLevel={1}
        zoomMode="fit"
      />
    </WithStyles(Drawer)>
    <Connect(PageListComponent)
      activePage={1}
      elementNamePrefix="Page-"
      fitRelativeZoomLevel={1}
      handleMarkerCreation={[Function]}
      id="sn-document-viewer-pages"
      images="preview"
      isPlacingCommentMarker={false}
      onPageClick={[Function]}
      padding={8}
      showWidgets={true}
      tolerance={0}
      zoomLevel={1}
      zoomMode="custom"
    />
    <WithStyles(Drawer)
      PaperProps={
        Object {
          "style": Object {
            "height": "100%",
            "overflow": "hidden",
            "position": "relative",
            "width": 0,
          },
        }
      }
      anchor="right"
      open={false}
      variant="persistent"
    >
      <styled.div>
        <WithStyles(Typography)
          variant="h4"
        />
        <CreateComment
          createComment={[Function]}
          handlePlaceMarkerClick={[Function]}
          isPlacingMarker={false}
          localization={Object {}}
        />
      </styled.div>
    </WithStyles(Drawer)>
  </div>
</div>
`;

exports[`Document Viewer Layout component should render without crashing 2`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "width": "100%",
    }
  }
>
  some children
  <div
    style={
      Object {
        "display": "flex",
        "height": "100%",
        "overflow": "hidden",
        "position": "relative",
        "width": "100%",
        "zIndex": 0,
      }
    }
  >
    <WithStyles(Drawer)
      PaperProps={
        Object {
          "style": Object {
            "height": "100%",
            "overflow": "hidden",
            "position": "relative",
            "width": 0,
          },
        }
      }
      anchor="left"
      open={false}
      variant="persistent"
    >
      <Connect(PageListComponent)
        activePage={1}
        elementNamePrefix="Thumbnail-"
        fitRelativeZoomLevel={0}
        id="sn-document-viewer-thumbnails"
        images="thumbnail"
        onPageClick={[Function]}
        padding={16}
        showWidgets={false}
        style={
          Object {
            "marginRight": "-16px",
            "minWidth": 200,
            "paddingRight": 0,
          }
        }
        tolerance={0}
        zoomLevel={1}
        zoomMode="fit"
      />
    </WithStyles(Drawer)>
    <Connect(PageListComponent)
      activePage={1}
      elementNamePrefix="Page-"
      fitRelativeZoomLevel={1}
      handleMarkerCreation={[Function]}
      id="sn-document-viewer-pages"
      images="preview"
      isPlacingCommentMarker={false}
      onPageClick={[Function]}
      padding={8}
      showWidgets={true}
      tolerance={0}
      zoomLevel={1}
      zoomMode="custom"
    />
    <WithStyles(Drawer)
      PaperProps={
        Object {
          "style": Object {
            "height": "100%",
            "overflow": "hidden",
            "position": "relative",
            "width": 0,
          },
        }
      }
      anchor="right"
      open={false}
      variant="persistent"
    >
      <styled.div>
        <WithStyles(Typography)
          variant="h4"
        />
        <CreateComment
          createComment={[Function]}
          handlePlaceMarkerClick={[Function]}
          isPlacingMarker={false}
          localization={Object {}}
        />
      </styled.div>
    </WithStyles(Drawer)>
  </div>
</div>
`;
